File: server/internal/accounts/db/plaid_items.go
package accountsdb

import (
	"context"

	"tallyo/internal/accounts"
	"tallyo/internal/database/dbgen"
	"tallyo/internal/database/dbutil"
	"tallyo/internal/graph/model"
)

func (s *Store) UpsertPlaidItem(ctx context.Context, item accounts.PlaidItemSecret) (int64, error) {
	id, err := s.q.UpsertPlaidItem(ctx, dbgen.UpsertPlaidItemParams{
		ExternalID:              item.ExternalID,
		CredentialID:            int64(item.CredentialID),
		AccessToken:             item.AccessToken,
		InstitutionID:           item.InstitutionID,
		LogoUrl:                 item.LogoURL,
		NextSyncAt:              dbutil.UTCTimePtr(item.NextSyncAt),
		NextRecurringSyncAt:     dbutil.UTCTimePtr(item.NextRecurringSyncAt),
		NextBalanceSyncAt:       dbutil.UTCTimePtr(item.NextBalanceSyncAt),
		PlaidInvestmentsEnabled: item.InvestmentsEnabled,
		PlaidLiabilitiesEnabled: item.LiabilitiesEnabled,
	})
	if err != nil {
		return 0, err
	}
	return id, nil
}

func (s *Store) PlaidItems(ctx context.Context, includeInactive bool) ([]*model.PlaidItem, error) {
	rows, err := s.q.ListPlaidItems(ctx, dbgen.ListPlaidItemsParams{ActiveOnly: !includeInactive})
	return dbutil.MapRows(rows, err, PlaidItemFromSQLRow)
}

func (s *Store) PlaidItem(ctx context.Context, id int64) (*model.PlaidItem, error) {
	items, err := s.PlaidItemsByIDs(ctx, []int64{id})
	return dbutil.MapSingle(items, err, id)
}

func (s *Store) PlaidItemsByIDs(ctx context.Context, ids []int64) (map[int64]*model.PlaidItem, error) {
	if len(ids) == 0 {
		return map[int64]*model.PlaidItem{}, nil
	}
	rows, err := s.q.ListPlaidItems(ctx, dbgen.ListPlaidItemsParams{Ids: ids})
	toPlaidItemByID := func(row dbgen.ListPlaidItemsRow) (int64, *model.PlaidItem) {
		return row.ItemID, PlaidItemFromSQLRow(row)
	}
	return dbutil.AssociateRows(rows, err, toPlaidItemByID)
}

func PlaidItemFromSQLRow(row dbgen.ListPlaidItemsRow) *model.PlaidItem {
	item := &model.PlaidItem{
		ID:                  model.New(model.GlobalIDPlaidItem, row.ItemID),
		Credential:          &model.PlaidCredential{ID: int32(row.CredentialID)},
		InstitutionID:       row.InstitutionID,
		LastSyncedAt:        row.LastSyncedAt,
		HealthState:         model.PlaidItemHealthState(row.HealthState),
		HealthErrorCode:     row.HealthErrorCode,
		HealthErrorMessage:  row.HealthErrorMessage,
		HealthUpdatedAt:     row.HealthUpdatedAt,
		SyncCron:            row.SyncCron,
		RecurringSyncCron:   row.RecurringSyncCron,
		NextSyncAt:          row.NextSyncAt,
		NextRecurringSyncAt: row.NextRecurringSyncAt,
		IsActive:            row.ConnectionIsActive,
		CreatedAt:           row.CreatedAt,
		UpdatedAt:           row.UpdatedAt,
	}
	if !item.HealthState.IsValid() {
		item.HealthState = model.PlaidItemHealthStateHealthy
	}
	return item
}

func (s *Store) PlaidItemSecret(ctx context.Context, id int64) (*accounts.PlaidItemSecret, error) {
	return PlaidItemSecretByID(ctx, s.q, id, accounts.PlaidSyncKindTransactions)
}

func PlaidItemSecretByID(ctx context.Context, q *dbgen.Queries, id int64, kind accounts.PlaidSyncKind) (*accounts.PlaidItemSecret, error) {
	rows, err := q.PlaidItemsDue(ctx, dbgen.PlaidItemsDueParams{ID: &id, Kind: string(kind)})
	toSecret := func(row dbgen.PlaidItemsDueRow) *accounts.PlaidItemSecret {
		item := PlaidItemSecretFromSQLRow(row)
		return &item
	}
	return dbutil.MapFirstRow(rows, err, toSecret)
}

func PlaidItemSecretFromSQLRow(row dbgen.PlaidItemsDueRow) accounts.PlaidItemSecret {
	item := row.PlaidItem
	return accounts.PlaidItemSecret{
		ID:                  item.ID,
		ExternalID:          item.ExternalID,
		CredentialID:        int32(item.CredentialID),
		AccessToken:         item.AccessToken,
		OwnerID:             row.OwnerID,
		OwnerName:           row.Owner,
		InstitutionID:       item.InstitutionID,
		InstitutionName:     row.InstitutionName,
		LogoURL:             item.LogoUrl,
		LastSyncedAt:        item.LastSyncedAt,
		SyncCron:            item.SyncCron,
		RecurringSyncCron:   item.RecurringSyncCron,
		NextSyncAt:          item.NextSyncAt,
		NextRecurringSyncAt: item.NextRecurringSyncAt,
		NextBalanceSyncAt:   item.NextBalanceSyncAt,
		InvestmentsEnabled:  item.PlaidInvestmentsEnabled,
		LiabilitiesEnabled:  item.PlaidLiabilitiesEnabled,
	}
}

func (s *Store) PlaidItemsMissingLogoURL(ctx context.Context) ([]accounts.PlaidItemSecret, error) {
	rows, err := s.q.PlaidItemsMissingLogoURL(ctx)
	fromRow := func(row dbgen.PlaidItemsMissingLogoURLRow) accounts.PlaidItemSecret {
		return accounts.PlaidItemSecret{
			ID:            row.ID,
			ExternalID:    row.ExternalID,
			CredentialID:  int32(row.CredentialID),
			InstitutionID: row.InstitutionID,
		}
	}
	return dbutil.MapRows(rows, err, fromRow)
}

func (s *Store) SetPlaidItemLogoURL(ctx context.Context, itemID int64, logoURL *string) error {
	return s.q.SetPlaidItemLogoURL(ctx, dbgen.SetPlaidItemLogoURLParams{LogoUrl: logoURL, ID: itemID})
}

func (s *Store) ResetItemSync(ctx context.Context, itemID int64) error {
	return s.q.ResetItemSync(ctx, dbgen.ResetItemSyncParams{ID: itemID})
}

func (s *Store) SetPlaidProductFlags(ctx context.Context, itemID int64, investmentsEnabled bool, liabilitiesEnabled bool) error {
	return s.q.SetPlaidProductFlags(ctx, dbgen.SetPlaidProductFlagsParams{
		ID:                      itemID,
		PlaidInvestmentsEnabled: investmentsEnabled,
		PlaidLiabilitiesEnabled: liabilitiesEnabled,
	})
}

File: server-rs/src/accounts/store/plaid_items.rs
use std::collections::HashMap;

use anyhow::Result;
use chrono::{DateTime, Utc};
use sqlx::{Executor, Sqlite};

use crate::{
    accounts::{PlaidItemHealthState, PlaidItemRecord, PlaidItemSecret, PlaidSyncKind},
    database::queries,
};

pub async fn upsert_plaid_item(
    executor: impl Executor<'_, Database = Sqlite>,
    params: queries::UpsertPlaidItemParams<'_>,
) -> Result<i64> {
    queries::upsert_plaid_item(executor, params)
        .await
        .map(|row| row.id)
        .map_err(Into::into)
}

pub async fn plaid_items(
    executor: impl Executor<'_, Database = Sqlite>,
    include_inactive: bool,
) -> Result<Vec<PlaidItemRecord>> {
    queries::list_plaid_items(
        executor,
        queries::ListPlaidItemsParams {
            active_only: !include_inactive,
            ..Default::default()
        },
    )
    .await
    .map(|rows| rows.into_iter().map(Into::into).collect())
    .map_err(Into::into)
}

pub async fn plaid_item_by_id(
    executor: impl Executor<'_, Database = Sqlite>,
    id: i64,
) -> Result<Option<PlaidItemRecord>> {
    let ids = [id];
    queries::list_plaid_items(
        executor,
        queries::ListPlaidItemsParams {
            ids: Some(&ids),
            ..Default::default()
        },
    )
    .await
    .map(|rows| rows.into_iter().next().map(Into::into))
    .map_err(Into::into)
}

pub async fn plaid_items_by_ids(
    executor: impl Executor<'_, Database = Sqlite>,
    ids: &[i64],
) -> Result<HashMap<i64, PlaidItemRecord>> {
    if ids.is_empty() {
        return Ok(HashMap::new());
    }
    queries::list_plaid_items(
        executor,
        queries::ListPlaidItemsParams {
            ids: Some(ids),
            ..Default::default()
        },
    )
    .await
    .map(|rows| {
        rows.into_iter()
            .map(PlaidItemRecord::from)
            .map(|record| (record.item.id, record))
            .collect()
    })
    .map_err(Into::into)
}

pub async fn plaid_item_secret_by_id(
    executor: impl Executor<'_, Database = Sqlite>,
    id: i64,
    kind: PlaidSyncKind,
) -> Result<Option<PlaidItemSecret>> {
    queries::plaid_items_due(
        executor,
        queries::PlaidItemsDueParams {
            id: Some(id),
            kind: &kind.to_string(),
            due_only: false,
            ..Default::default()
        },
    )
    .await
    .map(|rows| rows.into_iter().next())
    .map_err(Into::into)
}

pub async fn plaid_items_due(
    executor: impl Executor<'_, Database = Sqlite>,
    kind: PlaidSyncKind,
    now: DateTime<Utc>,
) -> Result<Vec<PlaidItemSecret>> {
    queries::plaid_items_due(
        executor,
        queries::PlaidItemsDueParams {
            kind: &kind.to_string(),
            now: Some(now.into()),
            due_only: true,
            ..Default::default()
        },
    )
    .await
    .map_err(Into::into)
}

pub async fn reset_item_sync(executor: impl Executor<'_, Database = Sqlite>, item_id: i64) -> Result<()> {
    queries::reset_item_sync(executor, queries::ResetItemSyncParams { id: item_id })
        .await
        .map_err(Into::into)
}

pub async fn set_plaid_product_flags(
    executor: impl Executor<'_, Database = Sqlite>,
    item_id: i64,
    investments_enabled: bool,
    liabilities_enabled: bool,
) -> Result<()> {
    queries::set_plaid_product_flags(
        executor,
        queries::SetPlaidProductFlagsParams {
            plaid_investments_enabled: investments_enabled,
            plaid_liabilities_enabled: liabilities_enabled,
            id: item_id,
        },
    )
    .await
    .map_err(Into::into)
}

pub async fn set_plaid_item_health(
    executor: impl Executor<'_, Database = Sqlite>,
    item_id: i64,
    health_state: PlaidItemHealthState,
    error_code: Option<&str>,
    error_message: Option<&str>,
) -> Result<()> {
    queries::set_plaid_item_health(
        executor,
        queries::SetPlaidItemHealthParams {
            health_state: &health_state.to_string(),
            health_error_code: error_code,
            health_error_message: error_message,
            id: item_id,
        },
    )
    .await
    .map_err(Into::into)
}

pub async fn set_item_balance_synced(
    executor: impl Executor<'_, Database = Sqlite>,
    item_id: i64,
    next_balance_sync_at: DateTime<Utc>,
) -> Result<()> {
    queries::set_item_balance_synced(
        executor,
        queries::SetItemBalanceSyncedParams {
            next_balance_sync_at: Some(next_balance_sync_at.into()),
            id: item_id,
        },
    )
    .await
    .map_err(Into::into)
}

#[cfg(test)]
mod tests {
    use anyhow::Result;
    use chrono::{DateTime, Utc};

    use crate::{
        database::{Timestamp, dbtest, queries},
        testutil::store::{create_owner, create_plaid_credential},
    };

    use super::*;

    #[tokio::test]
    async fn reused_plaid_item_secret_round_trips_storage_fields() -> Result<()> {
        let pool = dbtest::open().await?;
        let owner = create_owner(&pool, "alex").await?;
        let credential_id = create_plaid_credential(&pool).await?;
        let timestamp = "2026-09-06T12:00:00Z".parse::<DateTime<Utc>>()?;
        let item_id = upsert_plaid_item(
            &pool,
            queries::UpsertPlaidItemParams {
                external_id: "plaid-item",
                credential_id,
                access_token: "access-token",
                institution_id: Some("institution"),
                logo_url: Some("https://bank.example/logo"),
                next_sync_at: Some(timestamp.into()),
                next_recurring_sync_at: Some(timestamp.into()),
                next_balance_sync_at: Some(timestamp.into()),
                plaid_investments_enabled: true,
                plaid_liabilities_enabled: true,
            },
        )
        .await?;
        super::super::create_connection(&pool, item_id, Some("Bank"), owner.id).await?;
        sqlx::query(
            "UPDATE plaid_items SET cursor = ?, health_state = ?, health_error_code = ?, health_error_message = ?, health_updated_at = ?, last_recurring_synced_at = ?, last_synced_at = ?, sync_cron = ?, recurring_sync_cron = ?, created_at = ?, updated_at = ? WHERE id = ?",
        )
        .bind("cursor")
        .bind("SYNC_ERROR")
        .bind("code")
        .bind("message")
        .bind(Timestamp::from(timestamp))
        .bind(Timestamp::from(timestamp))
        .bind(Timestamp::from(timestamp))
        .bind("0 1 * * *")
        .bind("0 2 * * *")
        .bind(Timestamp::from(timestamp))
        .bind(Timestamp::from(timestamp))
        .bind(item_id)
        .execute(&pool)
        .await?;

        let secret = plaid_item_secret_by_id(&pool, item_id, crate::accounts::PlaidSyncKind::Sync)
            .await?
            .unwrap();
        assert_eq!(secret.owner_id, owner.id);
        assert_eq!(secret.owner, owner.name);
        assert_eq!(secret.institution_name.as_deref(), Some("Bank"));
        assert_eq!(secret.plaid_items.id, item_id);
        assert_eq!(secret.plaid_items.external_id, "plaid-item");
        assert_eq!(secret.plaid_items.credential_id, credential_id);
        assert_eq!(secret.plaid_items.access_token, "access-token");
        assert_eq!(secret.plaid_items.institution_id.as_deref(), Some("institution"));
        assert_eq!(
            secret.plaid_items.logo_url.as_deref(),
            Some("https://bank.example/logo")
        );
        assert_eq!(secret.plaid_items.cursor.as_deref(), Some("cursor"));
        assert_eq!(secret.plaid_items.health_state, "SYNC_ERROR");
        assert_eq!(secret.plaid_items.health_error_code.as_deref(), Some("code"));
        assert_eq!(secret.plaid_items.health_error_message.as_deref(), Some("message"));
        assert_eq!(secret.plaid_items.health_updated_at, Some(timestamp.into()));
        assert_eq!(secret.plaid_items.last_recurring_synced_at, Some(timestamp.into()));
        assert_eq!(secret.last_synced_at(), Some(timestamp));
        assert_eq!(secret.plaid_items.sync_cron, "0 1 * * *");
        assert_eq!(secret.plaid_items.recurring_sync_cron, "0 2 * * *");
        assert_eq!(secret.next_sync_at(), Some(timestamp));
        assert_eq!(secret.next_recurring_sync_at(), Some(timestamp));
        assert_eq!(secret.next_balance_sync_at(), Some(timestamp));
        assert!(secret.plaid_items.plaid_investments_enabled);
        assert!(secret.plaid_items.plaid_liabilities_enabled);
        assert_eq!(secret.plaid_items.created_at, timestamp.into());
        assert_eq!(secret.plaid_items.updated_at, timestamp.into());
        Ok(())
    }

    #[tokio::test]
    async fn empty_batch_lookups_return_empty_maps() -> Result<()> {
        let pool = dbtest::open().await?;
        assert!(plaid_items_by_ids(&pool, &[]).await?.is_empty());
        Ok(())
    }
}
